src/WebJobs.Script/Host/ScriptHostManager.cs

```diff
@@ -44,16 +47,18 @@ public class ScriptHostManager : IScriptHostEnvironment, IDisposable
         private ILogger _logger;
 
         private ScriptSettingsManager _settingsManager;
-        private AutoRecoveringFileSystemWatcher _scriptFileWatcher;
         private CancellationTokenSource _restartDelayTokenSource;
 
         public ScriptHostManager(ScriptHostConfiguration config, IScriptEventManager eventManager = null, IScriptHostEnvironment environment = null)
             : this(config, ScriptSettingsManager.Instance, new ScriptHostFactory(), eventManager, environment)
         {
             if (config.FileWatchingEnabled)
             {
-                _scriptFileWatcher = new AutoRecoveringFileSystemWatcher(config.RootScriptPath);
-                _scriptFileWatcher.Changed += OnScriptFileChanged;
+                _fileEventSource = new FileWatcherEventSource(EventManager, "ScriptFiles", config.RootScriptPath);
+
+                EventManager.OfType<FileEvent>()
+                    .Where(f => string.Equals(f.Source, "ScriptFiles", StringComparison.Ordinal))
```

Comment:
Would it be useful to add a Linq extension like `WithSource(string)` since this looks like it'll be the common pattern? This is assuming `Source` moves to `ScriptEvent`, which may or may not make sense.